File: backend-ts/src/db/sqlitePath.mjs
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

/** `backend-ts/` (this package root). */
const packageRoot = path.join(__dirname, '..', '..');
/** Monorepo root — sibling of `backend-ts`; holds `docs/`, `store/`. */
export const repoRoot = path.join(packageRoot, '..');

const CONFIG_IN_PACKAGE = path.join(packageRoot, 'config', 'backend.yaml');
const SCHEMA_RELATIVE = path.join('docs', 'sqlite', 'schema.sql');

const DEFAULT_DB_RELATIVE = path.join('store', 'app.sqlite');

/**
 * Read `database.path` from `backend-ts/config/backend.yaml`. No YAML dependency — scalar `path` under `database:` only.
 * @returns {string | null} path as written in YAML (often relative to package root), or null if unset / unreadable
 */
export function readDatabasePathFromPackageConfig() {
  try {
    const raw = fs.readFileSync(CONFIG_IN_PACKAGE, 'utf8');
    return parseDatabasePathYaml(raw);
  } catch {
    return null;
  }
}

/**
 * @param {string} content full `backend.yaml` text
 * @returns {string | null}
 */
export function parseDatabasePathYaml(content) {
  const lines = content.split(/\r?\n/);
  let inDatabase = false;
  for (const line of lines) {
    const trimmed = line.trim();
    if (trimmed === '' || trimmed.startsWith('#')) continue;
    if (/^database:\s*(#.*)?$/.test(trimmed)) {
      inDatabase = true;
      continue;
    }
    if (inDatabase) {
      if (/^\S/.test(line)) break;
      const quoted = trimmed.match(/^path:\s*["'](.+)["']\s*(?:#.*)?$/);
      if (quoted) return quoted[1].trim();
      const plain = trimmed.match(/^path:\s*([^\s#]+)/);
      if (plain) return plain[1].trim();
    }
  }
  return null;
}

/**
 * Absolute path to the SQLite file.
 * Precedence: **`SQLITE_PATH`** env → **`backend-ts/config/backend.yaml`** `database.path` (relative to `backend-ts/`) → `../store/app.sqlite` from package via default under repo root.
 */
export function resolveSqlitePath() {
  const fromEnv = process.env.SQLITE_PATH?.trim();
  if (fromEnv) return path.resolve(fromEnv);
  const fromYaml = readDatabasePathFromPackageConfig();
  if (fromYaml) return path.resolve(packageRoot, fromYaml);
  return path.resolve(repoRoot, DEFAULT_DB_RELATIVE);
}

/**
 * Load canonical DDL from `docs/sqlite/schema.sql` at monorepo root (single source of truth for Node + Go).
 */
export function loadCanonicalSchemaSql() {
  const schemaPath = path.join(repoRoot, SCHEMA_RELATIVE);
  if (!fs.existsSync(schemaPath)) {
    throw new Error(
      `[sqlite] Missing schema file: ${schemaPath}. Restore docs/sqlite/schema.sql (see docs/BACKEND_SHARED_CONTRACT.md).`
    );
  }
  return fs.readFileSync(schemaPath, 'utf8');
}

File: backend-go/pkg/sqlitepath/sqlitepath.go
package sqlitepath

import (
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

const defaultRelative = "store" + string(filepath.Separator) + "app.sqlite"

// Resolve returns the absolute path to the SQLite file.
// Order: SQLITE_PATH env → backend-go/config/backend.yaml `database.path` (relative to packageRoot) → repoRoot/store/app.sqlite.
func Resolve(packageRoot, repoRoot string) (string, error) {
	if p := strings.TrimSpace(os.Getenv("SQLITE_PATH")); p != "" {
		if filepath.IsAbs(p) {
			return filepath.Clean(p), nil
		}
		return filepath.Abs(p)
	}
	cfgPath := filepath.Join(packageRoot, "config", "backend.yaml")
	rel := readDatabasePathFromYAML(cfgPath)
	if rel == "" {
		return filepath.Join(repoRoot, filepath.FromSlash("store/app.sqlite")), nil
	}
	rel = filepath.FromSlash(strings.TrimSpace(rel))
	if filepath.IsAbs(rel) {
		return filepath.Clean(rel), nil
	}
	return filepath.Abs(filepath.Join(packageRoot, rel))
}

func readDatabasePathFromYAML(path string) string {
	raw, err := os.ReadFile(path)
	if err != nil {
		return ""
	}
	return parseDatabasePathYAML(string(raw))
}

var (
	reDatabaseLine = regexp.MustCompile(`^database:\s*(#.*)?$`)
	rePathQuoted   = regexp.MustCompile(`^path:\s*["'](.+)["']\s*(?:#.*)?$`)
	rePathPlain    = regexp.MustCompile(`^path:\s*([^\s#]+)`)
)

func parseDatabasePathYAML(content string) string {
	lines := strings.Split(strings.ReplaceAll(content, "\r\n", "\n"), "\n")
	inDatabase := false
	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		if trimmed == "" || strings.HasPrefix(trimmed, "#") {
			continue
		}
		if reDatabaseLine.MatchString(trimmed) {
			inDatabase = true
			continue
		}
		if inDatabase {
			if len(line) > 0 && !strings.ContainsAny(string(line[0]), " \t") {
				break
			}
			if m := rePathQuoted.FindStringSubmatch(trimmed); len(m) > 1 {
				return strings.TrimSpace(m[1])
			}
			if m := rePathPlain.FindStringSubmatch(trimmed); len(m) > 1 {
				return strings.TrimSpace(m[1])
			}
		}
	}
	return ""
}

// DefaultRelativePath documents the fallback under repo root (for README).
func DefaultRelativePath() string { return defaultRelative }
